packages/editor/src/components/rich-text/list-edit.js

```diff
@@ -13,60 +18,105 @@ import { Fragment } from '@wordpress/element';
 import { RichTextShortcut } from './shortcut';
 import BlockFormatControls from '../block-format-controls';
 
-function isListRootSelected( editor ) {
-	return (
-		! editor.selection ||
-		editor.selection.getNode().closest( 'ol,ul' ) === editor.getBody()
-	);
-}
+const { TEXT_NODE, ELEMENT_NODE } = window.Node;
 
-function isActiveListType( editor, tagName, rootTagName ) {
-	if ( document.activeElement !== editor.getBody() ) {
-		return tagName === rootTagName;
+/**
+ * Gets the selected list node, which is the closest list node to the start of
+ * the selection.
+ *
+ * @return {?Element} The selected list node, or undefined if none is selected.
+ */
+function getSelectedListNode() {
+	const selection = window.getSelection();
+
+	if ( selection.rangeCount === 0 ) {
+		return;
 	}
 
-	const listItem = editor.selection.getNode();
-	const list = listItem.closest( 'ol,ul' );
+	let { startContainer } = selection.getRangeAt( 0 );
+
+	if ( startContainer.nodeType === TEXT_NODE ) {
+		startContainer = startContainer.parentNode;
+	}
 
-	if ( ! list ) {
+	if ( startContainer.nodeType !== ELEMENT_NODE ) {
 		return;
 	}
 
-	return list.nodeName.toLowerCase() === tagName;
+	const rootNode = startContainer.closest( '*[contenteditable]' );
+
+	if ( ! rootNode || ! rootNode.contains( startContainer ) ) {
+		return;
+	}
+
+	return startContainer.closest( 'ol,ul' );
 }
 
-export const ListEdit = ( { editor, onTagNameChange, tagName, onSyncDOM } ) => (
+/**
+ * Whether or not the root list is selected.
+ *
+ * @return {boolean} True if the root list or nothing is selected, false if an
+ *                   inner list is selected.
+ */
+function isListRootSelected() {
+	const listNode = getSelectedListNode();
+
+	// Consider the root list selected if nothing is selected.
+	return ! listNode || listNode.contentEditable === 'true';
+}
+
+/**
+ * Wether or not the selected list has the given tag name.
+ *
+ * @param {string}  tagName     The tag name the list should have.
+ * @param {string}  rootTagName The current root tag name, to compare with in
+ *                              case nothing is selected.
+ *
+ * @return {boolean}             [description]
+ */
+function isActiveListType( tagName, rootTagName ) {
+	const listNode = getSelectedListNode();
+
+	if ( ! listNode ) {
+		return tagName === rootTagName;
+	}
+
+	return listNode.nodeName.toLowerCase() === tagName;
+}
+
+export const ListEdit = ( {
+	onTagNameChange,
+	tagName,
+	value,
+	onChange,
+} ) => (
```

Comment:
Slightly tangential to the PR, but I would've expected the relationship between RichText and the List block to be more like:

```
List#edit = ( … ) =>
  <ListEdit>
    <RichText />
  </ListEdit>
```

rather than let `RichText` conditionally render `ListEdit` among its children based its own state and props (`isSelected && this.multilineTag === 'li'`). It feels like a strange burden on RichText and adds some indirection that reminds me of React/Backbone mixins, since `List#edit` has to remember to pass `onTagNameChange` to `RichText`. Also, I wonder why we need to generalise `onTagNameChange` instead of letting `ListEdit` implement specific logic for `ul` and `ol`.

Reply:
Sounds good to rethink, but I feel like it might be better to do separately because we leave most of this unchanged.

Reply:
Agreed